geopy/geocoders/osm.py

```diff
@@ -180,6 +181,15 @@ def geocode(
             'format': 'json'
         })
 
+        if limit:
+            limit = int(limit)
+            if limit < 1:
+                raise ValueError("Limit cannot be less than 1")
+            exactly_one = limit == 1
```

Comment:
`exactly_one` affects the return type of the function call (a single location vs a list of locations), so I believe that in case of conflicting `limit` and `exactly_one` values the latter should take precedence.

What do you think of removing this line and turning the following `elif` to a simple `if`, which would force `limit=1` for `exactly_one=True`?

Reply:
Just to be sure, is this what you have in mind ?

```python
if exactly_one:
  params['limit'] = 1
elif limit:
  limit = int(limit)
  if limit < 1:
    raise ValueError("Limit cannot be less than 1")
  params['limit'] = limit
```

If I write that, given that `exactly_one` is true by default, we will always have that `limit=1`. Also, the result will always be a single location result (and not a list, even if a limit is given). To change this (i.e. have a limit other than one **and** a list of results), we would have to explicitly set `exactly_one` to false in addition to passing a limit, which is a bit cumbersome in my opinion.

I initially wanted to raise an exception if both exactly_one and limit were set, but as said above such a case is not ideal. That's why I made it so that if limit is given, the value of exactly_one is changed and determined by `exactly_one = limit == 1`.

What are your thoughts on this ?

Reply:
I was trying to describe something like this:
```python
        if limit:
            limit = int(limit)
            if limit < 1:
                raise ValueError("Limit cannot be less than 1")
            params['limit'] = limit
        if exactly_one:
            params['limit'] = 1
```

Your snippet does the same (except that in your case a ValueError won't be thrown for `exactly_one=True, limit=-1`). Both are fine IMO, pick the one that you prefer.

> To change this (i.e. have a limit other than one and a list of results), we would have to explicitly set exactly_one to false in addition to passing a limit, which is a bit cumbersome in my opinion.

It's the current convention to explicitly set `exactly_one` to False for getting a list. I kinda agree that it looks cumbersome to specify both, but violating that convention doesn't seem reasonable to me.

`exactly_one` is implemented by all geocoders. Basically it means what kind of result does the user expect from this geocoding call. It's universal across all geocoder classes. `limit` is just an optional arg of this specific geocoder which allows to tell the backend a *desired* amount of results. But `limit` in no way controls the real number of locations in a result: it can be less, it can be more. It's up to the backend to decide.

So the idea is that `exactly_one` always takes precedence over `limit`.

Thereby I think there's no need to raise an exception if both are set – from my point of view it seems valid (except the `exactly_one=True` and `limit != 1` case, which can be silently fixed to `limit = 1` without being ambiguous to the API user, given that it's `exactly_one` arg which controls the response type rather than `limit`).

Does that seem reasonable to you?

Reply:
Put this way it is wiser to continue with the current convention you mentioned. As such, it should be made clear in the documentation how the limit parameter is handled.

Doing changes now.

Reply:
Done in 05651b5